airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/source/relationaldb/state/CursorStateMessageProducer.java

```diff
@@ -0,0 +1,144 @@
+          + "data loss can occur.");
+    }
+    return message;
+  }
+
+  /**
+   * Note: We do not try to catch exception here. If error/exception happens, we should fail the sync,
+   * and since we have saved state message before, we should be able to resume it in next sync if we
+   * have fixed the underlying issue, of if the issue is transient.
+   *
+   * @param message
+   * @return
+   */
+  @Override
+  public AirbyteMessage processRecordMessage(final ConfiguredAirbyteStream stream, AirbyteMessage message) {
+    totalRecordCount++;
+    final String cursorField = IncrementalUtils.getCursorField(stream);
+    if (message.getRecord().getData().hasNonNull(cursorField)) {
+      final String cursorCandidate = getCursorCandidate(cursorField, message);
+      final JsonSchemaPrimitive cursorType = IncrementalUtils.getCursorType(stream,
+          cursorField);
+      final int cursorComparison = IncrementalUtils.compareCursors(currentMaxCursor, cursorCandidate, cursorType);
+      if (cursorComparison < 0) {
+        // Update the current max cursor only when current max cursor < cursor candidate from the message
+        if (!Objects.equals(currentMaxCursor, initialCursor)) {
+          // Only create an intermediate state when it is not the first record.
+          intermediateStateMessage = createStateMessage(stream, totalRecordCount);
+        }
+        currentMaxCursor = cursorCandidate;
+        currentMaxCursorRecordCount = 1L;
+      } else if (cursorComparison == 0) {
+        currentMaxCursorRecordCount++;
+      } else if (cursorComparison > 0) {
+        cursorOutOrOrderDetected = true;
+      }
+    }
+    return message;
+
+  }
+
+  @Override
+  public AirbyteStateMessage createFinalStateMessage(final ConfiguredAirbyteStream stream) {
+    return createStateMessage(stream, totalRecordCount);
+  }
+
+  /**
+   * Only sends out state message when there is a state message to be sent out.
+   */
+  @Override
+  public boolean shouldEmitStateMessage(final ConfiguredAirbyteStream stream) {
+    return intermediateStateMessage != null;
+  }
+
+  /**
+   * Creates AirbyteStateMessage while updating the cursor used to checkpoint the state of records
+   * read up so far
+   *
+   * @param recordCount count of total read messages. Used to determine log frequency.
+   * @return AirbyteMessage which includes information on state of records read so far
+   */
+  private AirbyteStateMessage createStateMessage(final ConfiguredAirbyteStream stream, final int recordCount) {
+    final AirbyteStreamNameNamespacePair pair = new AirbyteStreamNameNamespacePair(stream.getStream().getName(), stream.getStream().getNamespace());
+    final AirbyteStateMessage stateMessage = stateManager.updateAndEmit(pair, currentMaxCursor, currentMaxCursorRecordCount);
+    final Optional<CursorInfo> cursorInfo = stateManager.getCursorInfo(pair);
+
+    // logging once every 100 messages to reduce log verbosity
+    if (recordCount % 100 == 0) {
```

Comment:
Is 100 enough? Seems low. In any case, extract this magic number into a static variable.

Reply:
done. number 100 comes from the existing value in StateDecoratingIterator.